perf(ast/estree): skip escaping identifiers

overlookmotel opened the pull request:
Similar to #9396. JS identifiers cannot contain any of the characters which require escaping in JSON (whitespace, control characters, `"`, `\`).

Mark struct fields containing a JS identifier name with `#[estree(json_safe)]` to cause codegen to wrap them in `JsonSafeString` before serializing. This removes the work of escaping identifiers during serialization.

Note: Identifiers *can* contain `\` in source text as start of an escape sequence. But such escape sequences are unescaped in parser, so the `Atom` in AST representing identifier name cannot contain `\`.

overlookmotel commented:
* **#9398** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9398?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9398?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9398).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Feb 26, 9:46 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(ast/estree): skip escaping identifiers (#9398)

Similar to #9396. JS identifiers cannot contain any of the characters which require escaping in JSON (whitespace, control characters, `"`, `\`).

Mark struct fields containing a JS identifier name with `#[estree(json_safe)]` to cause codegen to wrap them in `JsonSafeString` before serializing. This removes the work of escaping identifiers during serialization.

Note: Identifiers *can* contain `\` in source text as start of an escape sequence. But such escape sequences are unescaped in parser, so the `Atom` in AST representing identifier name cannot contain `\`.
